### ISSUE TITLE
Java 9 Immutable List Nullpointer

### ISSUE THREAD

**Issue by 28Smiles** [jdbi/jdbi#1263]
public final This defineList(String key, List<?> values) {
            if (values.isEmpty()) {
                throw new IllegalArgumentException(
                        getClass().getSimpleName() + ".defineList was called with an empty list.");
            }
            if (values.contains(null)) {
                throw new IllegalArgumentException(
                        getClass().getSimpleName() + ".defineList was called with a list containing null values.");
            }
    
            String value = values.stream()
                    .map(Object::toString)
                    .collect(joining(", "));
    
            return define(key, value);
    }

`values.contains(null)` throws a nullpointer exception if list is created with `List.of()`, since it's `contains` method produces a NullpointerException if filled with null:

        @Override
        public boolean contains(Object o) {
            for (E e : elements) {
                if (o.equals(e)) { // implicit nullcheck of o
                    return true;
                }
            }
            return false;
        }

**Comment by qualidafial** [comment]
List.of() API docs say that NPE will be thrown for any null elements: https://docs.oracle.com/javase/9/docs/api/java/util/List.html#of-E-

**Comment by 28Smiles** [comment]
@qualidafial yes thats right, but in case of contains, the List will throw a NPE if asked if it contains null

**Comment by qualidafial** [comment]
Do you want Jdbi to stop calling contains() because one implementation doesn't guard against null?

**Comment by TheRealMarnes** [comment]
This very much sounds like a non-jdbi-issue to me, for what my opinion's worth. It's not jdbi's fault the list decides to throw an npe on null items it contains (why can you even get a null into it if it doesn't like nulls?). Jdbi needs to check for nulls, so it does. Your issue is with the list implementation you decided to use.

**Comment by 28Smiles** [comment]
@TheRealMarnes @qualidafial I agree it isn't an jdbi issue, but isn't it desireable to at least support all types of list the JCL provides, a fix would be to do it simmilar to the `bindList` implementation, with a stream.

**Comment by TheRealMarnes** [comment]
This List.of that throws an npe on null items (that you somehow got into the list in the first place) and doesn't throw that npe when you stream the items (because logic) is seriously a jdk implementation?

I mean, jdbi kinda has to support it then, but I can't wrap my head around such a thing actually being in the jdk.

**Comment by 28Smiles** [comment]
@TheRealMarnes  Yes it is, that is why I opened this Issue.

**Comment by TheRealMarnes** [comment]
I'm in favor of a workaround in jdbi then, but it'll be with an angry look toward the jdk and a comment explaining that the relevant code in jdbi has no right to exist then ¯\\\_(ツ)\_/¯

**Comment by 28Smiles** [comment]
I will open a pull request then

**Comment by 28Smiles** [comment]
#1265

**Comment by stevenschlansker** [comment]
Thanks for tracking this down, everyone :)

**Comment by dhardtke** [comment]
@TheRealMarnes You got something wrong there: The `List.of` factory method creates immutable `List` instances that **do not** permit null elements.

The [Javadoc](https://docs.oracle.com/javase/7/docs/api/java/util/List.html#contains(java.lang.Object)) is clear on the behavior of `java.util.List#contains(java.lang.Object)`:

```Throws NullPointerException - if the specified element is null and this list does not permit null elements (optional)```

So in my opinion this *is* a concern on the side of Jdbi and not an issue in the JDK itself - the JDK team made this decision (for some reason) and everyone has to stick to it.

**Comment by qualidafial** [comment]
Yeah, I just noticed that the optional `throws` clause has been there since Java 5. :nauseated_face: 

Still think it breaks LSP but it's in the spec so we'll deal with it
